src/Build/Collections/PropertyDictionary.cs

```diff
@@ -551,19 +551,34 @@ internal void Enumerate(Action<string, string> keyValueCallback)
 
         internal IEnumerable<TResult> Filter<TResult>(Func<T, bool> filter, Func<T, TResult> selector)
         {
-            List<TResult> result = new();
             lock (_properties)
             {
-                foreach (T property in (ICollection<T>)_properties)
+                ICollection<T> propertiesCollection = (ICollection<T>)_properties;
+                List<TResult> result = new(propertiesCollection.Count);
+                if (_properties is RetrievableValuedEntryHashSet<T> hashSet)
```

Comment:
Please add a comment that RetrievableValuedEntryHashSet<T> is prefered here since it has custome Enumerator implementation returning struct. Otherwise this change might easily get reverted by possible future refactorings